### ISSUE TITLE
running many replicates of find_graphs?

### ISSUE THREAD

**Issue by laufran** [uqrmaie1/admixtools#72]
Hi there,

In the older version of `find_graphs`, `find_graphs_old`, I see there's an argument `numrep` that allows for multiple runs of the same parameter arguments through one function call. Just to confirm, there's no such option for the current version of `find_graphs` / `qpgraph`? If so, how would you all recommend running many replicates? And how many replicates would you recommend, given the findings in [Maier et al. 2023](https://elifesciences.org/articles/85492) that "models fitting the data as well as or better than the true one are common, and their topological diversity is in most cases so high that it precludes consensus inference of topology by analysis of multiple topologies"?

Best,
Lauren

**Comment by uqrmaie1** [comment]
Yes, that is correct, there is no equivalent of `numrep` in `find_graphs()`. It seems a bit clearer to me to explicitly call the function multiple times than to use an argument for that, and it's not difficult to do. Here are two ways to do that:

```
numrep = 3
reslist = list()
for(i in seq_len(numrep)) {
  reslist[[i]] = find_graphs(f2_blocks, ...)
}
```
The code above uses only standard R syntax and functions, and the results in `reslist` will be in a list of data frames. Alternatively, you could do it like this:

```
numrep = 3
res = map(seq_len(numrep), ~find_graphs(f2_blocks, ...)) %>% bind_rows(.id='rep')
```
This will give you a single data frame where the replicate number is indicated by the column `rep`.
You could then get the graph with the lowest score in each replicate like this:
```
res %>% slice_min(score, by = rep)
```

If you model a complex graph or let it run for many generations, each replicate could take a while to run. In that case it might be better to parallelize across replicates, for example by submitting one job per replicate on a compute cluster, or using the `furrr` or `doParallel` R packages.

> And how many replicates would you recommend?

It depends on a few factors. Initially I would start with a small handful of replicates that you can inspect manually to get a feel of what the results look like. Later on, you might want to increase the number of replicates, depending on what the initial results look like:

* One possible outcome is that almost all replicates converge on the same best graph. In that case, there isn't really a point in running it too many times.
* Another possible outcome is that in every replicate you get a very different best graph, and all these best graphs have similarly low scores. In that case it also doesn't make sense to run it many times, since the results suggest that there isn't enough data to get anything meaningful, and you might want to reduce the complexity of the fitted model (reduce the number of admixture events).
* One outcome where it might make sense to run more replicates is if some of the replicates converge on a very good graph, but most replicates get stuck in a local optimum where the best graphs have poor scores. In that case, you might only find the global optimum graph after running many replicates.

Hope this helps!

**Comment by konopinski** [comment]
I've been trying to run repeated searches of the optimal tree but I encountered a weird problem. When I run the program in parallel I get the same outcome for every repeat. You may think - right, this is the ultimate tree and the search algorithm has to end up with it. The problem is that when just for curiosity I ran `find_graphs` independently even with smaller number of trees per generation, the search algorithm found a tree with lower score than in the ones from parallel run.
The script that uses multiple cores is:

```
f2blocks <- f2_from_geno("maxMiss099.recode.p.plink.AA", auto_only = FALSE)

cl <- makeForkCluster(35)
list_of_results <- parLapply(cl,1:35,function(i){
  nIter <- 5
  iteracja <- ((i-1) %% nIter)
  admix <- 1((i-1) %/% nIter)
  graph <- find_graphs(f2blocks, numadmix = admix, numgraphs = 50, stop_gen = 300, stop_gen2 = 30)
  graph <-  graph %>% slice_min(score, with_ties = FALSE)
  writeLines(as.character(graph$score),paste0("admixtools.score.",admix,".",iteracja,".txt"))
  pdf(paste0("admixtools.tree.",admix,".",iteracja,".pdf")); print(plot_graph(graph$edges[[1]])); dev.off()
  return(list(graph$score,graph$edges[[1]]))
})
stopCluster(cl)
```
and the one that provided a tree with lower score is just e.g.  
``` 
graph <- find_graphs(f2blocks, numadmix = 1, numgraphs = 20, stop_gen = 300, stop_gen2 = 30)
```
When I tried to run iterations being invoked by `for` loop the result was also same every time. Is it possible that the random seed is somehow set at the beginning of the execution of the code and the same one is used every time? If so it would be nice to have some control over it.

**Comment by uqrmaie1** [comment]
This seems to have more to do with `parallel` package than with `admixtools`.
You could try using `makeCluster` instead of `makeForkCluster`, which should give you different random seeds in each iteration:

```
> cl1 <- makeCluster(4)
> cl2 <- makeForkCluster(4)
> unlist(parLapply(cl1, 1:4, \(x) rnorm(1)))
[1] -0.6186925 -0.9418084  0.6255323 -0.4896532
> unlist(parLapply(cl2, 1:4, \(x) rnorm(1)))
[1] 0.6607697 0.6607697 0.6607697 0.6607697
```

`find_graphs` uses the `furrr` package to run things in parallel, and it uses the option `seed = TRUE` which should ensure that results are reproducible (you get the same result when re-running with the same initial random seed), and that randomly drawn numbers are not repeated across iterations (unless you fix the random seed in each iteration).

I'm not sure why you get identical results across iterations when you invoke them with a `for` loop. Could you share the code you used to do that?

**Comment by uqrmaie1** [comment]
Maybe I misunderstood your earlier post. It sounded like you get the same results for different iterations in a for loop without fixing the seed each time.

Regarding the optimal number of admixture edges: There are different opinions on this. My personal view is that it's better to stick with simple models with a very small number of allowed admixture events. It usually doesn't take more than 3 or 4 admixture events to get to a point where there is a large number of models that fit the data perfectly, and at that point it seems likely that you are just overfitting the data.
Most simple models are likely inaccurate, and large (absolute) log-likelihood scores and residual z-scores are indicators of that, but I don't think that any admixture graph model should be expected to be perfectly accurate. In most cases the actual demographic history is likely much more complex than what can be practically represented with admixture graphs.

The `qpgraph` scores are pseudo-log-likelihoods with the minus sign omitted, calculated as the weighted squared sum (or quadratic form) of f3 z-score residuals. The numbers should be negative, but I didn't want to depart from the original ADMIXTOOLS conventions.
